rust/src/storage/file/rename.rs

```diff
@@ -1,61 +1,81 @@
-        platform_specific_rename(from.as_ptr(), to.as_ptr())
-    };
 
-    if ret != 0 {
-        std::fs::remove_file(from).unwrap_or(());
+#[cfg(windows)]
+mod imp {
+    use super::*;
 
-        let e = errno::errno();
-        if let libc::EEXIST = e.0 {
-            return Err(StorageError::AlreadyExists(String::from(to)));
+    pub fn atomic_rename(from: &str, to: &str) -> Result<(), StorageError> {
+        // doing best effort in windows since there is no native atomic rename support
+        if std::fs::metadata(to).is_ok() {
+            return Err(StorageError::AlreadyExists(to.to_string()));
         }
+        std::fs::rename(from, to).map_err(|e| {
+            StorageError::other_std_io_err(format!("failed to rename {} to {}: {}", from, to, e))
+        })
+    }
+}
 
-        return Err(StorageError::Io {
-            source: custom_io_error(format!("{}", e)),
-        });
+#[cfg(unix)]
+mod imp {
+    use super::*;
+    use std::ffi::CString;
+
+    #[cfg(target_os = "linux")]
+    const RENAME_NOREPLACE: libc::c_uint = 1;
+
+    #[cfg(target_os = "linux")]
+    extern "C" {
+        fn renameat2(
+            olddirfd: libc::c_int,
+            oldpath: *const libc::c_char,
+            newdirfd: libc::c_int,
+            newpath: *const libc::c_char,
+            flags: libc::c_uint,
+        ) -> libc::c_int;
     }
 
-    Ok(())
-}
+    fn to_c_string(p: &str) -> Result<CString, StorageError> {
+        CString::new(p).map_err(|e| StorageError::Generic(format!("{}", e)))
+    }
+
+    pub fn atomic_rename(from: &str, to: &str) -> Result<(), StorageError> {
+        let ret = unsafe {
+            let from = to_c_string(from)?;
+            let to = to_c_string(to)?;
+            platform_specific_rename(from.as_ptr(), to.as_ptr())
+        };
+
+        if ret != 0 {
+            let e = errno::errno();
+            if let libc::EEXIST = e.0 {
+                return Err(StorageError::AlreadyExists(String::from(to)));
+            }
 
-unsafe fn platform_specific_rename(from: *const libc::c_char, to: *const libc::c_char) -> i32 {
-    cfg_if::cfg_if! {
-        if #[cfg(target_os = "linux")] {
-            renameat2(libc::AT_FDCWD, from, libc::AT_FDCWD, to, RENAME_NOREPLACE)
-        } else if #[cfg(target_os = "macos")] {
-            libc::renamex_np(from, to, libc::RENAME_EXCL)
-        } else {
-            unimplemented!()
+            return Err(StorageError::other_std_io_err(format!(
+                "failed to rename {} to {}: {}",
+                from, to, e
+            )));
         }
+
+        Ok(())
     }
-}
 
-fn to_c_string(p: &str) -> Result<CString, StorageError> {
-    CString::new(p).map_err(|e| StorageError::Generic(format!("{}", e)))
+    unsafe fn platform_specific_rename(from: *const libc::c_char, to: *const libc::c_char) -> i32 {
+        cfg_if::cfg_if! {
+            if #[cfg(target_os = "linux")] {
+                renameat2(libc::AT_FDCWD, from, libc::AT_FDCWD, to, RENAME_NOREPLACE)
```

Comment:
Reading the manpage for this function this little bit jumped out to me

> RENAME_NOREPLACE **requires support from the underlying filesystem.**

Emphasis mine of course. I don't have a good alternative suggestion, but just would like to express my worry that there couplings to underlying system implementation details that we cannot clearly communicate to users here.

Reply:
good call, added rust doc for filesystem back to clarify that.